R/helpers.R

```diff
@@ -361,21 +368,65 @@ LoadH5AD <- function(path) {
   }
   if (Exists(name = 'raw/X')) {
     md <- 'raw/var'
-    counts <- as.matrix(x = adata[['raw/X']])
+    x <- adata[['raw/X']]
   } else if (Exists(name = 'X')) {
     md <- 'var'
-    counts <- as.matrix(x = adata[['X']])
+    x <- adata[['X']]
   } else {
-    stop("Cannot find counts matrix")
+    stop("Cannot find counts matrix", call. = FALSE)
   }
-  if (!IsMetaData(md = md)) {
-    stop("Cannot find feature-level metadata", call. = FALSE)
-  } else if (!IsMetaData(md = 'obs')) {
-    stop("Cannot find cell-level metadata", call. = FALSE)
+  # check different possible attributes to try and get matrix shape
+  if (isTRUE(x$attr_exists(attr_name = 'h5sparse_shape'))) {
+    mtx.shape <- h5attr(x, 'h5sparse_shape')
+  } else if (isTRUE(x$attr_exists(attr_name = 'shape'))) {
+    mtx.shape <- h5attr(x, 'shape')
+  } else {
+    warning("Could not determine matrix shape")
+  }
+  # check different attributes to try and deduce matrix type
+  if (isTRUE(x$attr_exists(attr_name = 'h5sparse_format'))) {
+    mtx.type <- h5attr(x, 'h5sparse_format')
+  } else if (isTRUE(x$attr_exists(attr_name = 'encoding-type'))) {
+    mtx.type <- substr(h5attr(x, 'encoding-type'), 0, 3)
+  } else {
+    warning("Could not determine matrix format")
+  }
+  if (mtx.type != 'csr') {
+    p <- as.integer(x[['indptr']][])
+    i <- as.integer(x[['indices']][])
+    data <- as.double(x[['data']][])
+    # csc -> csr
+    converted.mtx <- csc_tocsr(
+      n_row = as.integer(mtx.shape[1]),
+      n_col = as.integer(mtx.shape[2]),
+      Ap = p,
+      Ai = i,
+      Ax = data
+    )
```

Comment:
Do we want to use the C++ code or use the other method where we read it in but don't reverse the dims?

Reply:
the cpp code was reliable and what I spent the most time testing with but I can experiment with the other approach for the next release since it's simpler